### ISSUE TITLE
Third-party transitive dependencies seem to be packaged into wheels incorrectly

### ISSUE THREAD

**Issue by jkaye2012** [bazelbuild/rules_python#2123]
# 🐞 bug report

### Affected Rule

py_wheel

### Is this a regression?

Doubtful

### Description

When building a wheel using `py_wheel`, third-party pip requirements are being installed into an incorrect directory, preventing them from being properly installed/used in Python environments.

## 🔬 Minimal Reproduction

```
load("@rules_python//python:defs.bzl", "py_binary", "py_library")
load("@rules_python//python:packaging.bzl", "py_package", "py_wheel")
load("@python_deps//:requirements.bzl", "requirement")

filegroup(
    name = "srcs",
    srcs = glob(
        [
            "**/*.py",
            "*.py",
        ],
        exclude = ["tests/**"],
    ),
)

py_deps = [
    requirement("behave"),
]

py_library(
    name = "regression_harness",
    srcs = [":srcs"],
    visibility = ["//bt:__subpackages__"],
    deps = py_deps,
)

py_package(
    name = "regression_harness_pkg",
    deps = [
        ":regression_harness",
    ],
)

py_wheel(
    name = "regression_harness_wheel",
    distribution = "regression_harness",
    version = "1.0.0",
    deps = [
        ":regression_harness_pkg",
    ],
)
```

## 🔥 Exception or Error

The wheel builds as usual, and looks (at least mostly) correct. However, upon installing the wheel, we see the following:

<pre><code>
gitpod /workspace/tribe $ ls /tmp/test/lib/python3.8/site-packages/
bin                       pip-24.1.2.virtualenv                  site-packages
bt                        pkg_resources                          _virtualenv.pth
_distutils_hack           __pycache__                            _virtualenv.py
distutils-precedence.pth  regression_harness-0.0.8940.dist-info  wheel
include                   setuptools                             wheel-0.43.0.dist-info
lib                       setuptools-57.4.0.dist-info            wheel-0.43.0.virtualenv
pip                       setuptools-57.4.0.virtualenv
pip-24.1.2.dist-info      share
</code></pre>

`regression_harness` was in fact installed, and is available, but the third-party transitive dependencies are within a **nested** site-packages directory _within_ the site-packages. 

<pre><code>
gitpod /workspace/tribe $ ls /tmp/test/lib/python3.8/site-packages/site-packages/
behave                  parse-1.20.2.dist-info  parse_type-0.6.2.dist-info  six-1.16.0.dist-info
behave-1.2.6.dist-info  parse.py                __pycache__                 six.py
__init__.py             parse_type              setuptools_behave.py
</code></pre>

Of course, this means that they're not available within the environment, as that directory is not part of the PYTHONPATH.

I have tried stripping the `site-packages` prefix when building the wheel, but this makes the wheel file invalid as it will contain multiple `.dist-info` files.

Am I doing something incorrectly? This seems like a bug to me, but I can't imagine that others wouldn't have seen this.

## 🌍  Your Environment

**Operating System:**

<pre>
  <code>
Ubuntu Focal
  </code>
</pre>

**Output of `bazel version`:**

<pre>
  <code>
gitpod /workspace/tribe $ bazel version
Bazelisk version: v1.17.0
Build label: 6.2.1
Build target: bazel-out/k8-opt/bin/src/main/java/com/google/devtools/build/lib/bazel/BazelServer_deploy.jar
Build time: Fri Jun 2 16:59:58 2023 (1685725198)
Build timestamp: 1685725198
Build timestamp as int: 1685725198
  </code>
</pre>

**Rules_python version:**

<pre>
  <code>
0.27.1
e85ae30de33625a63eca7fc40a94fea845e641888e52f32b6beea91e8b1b2793
  </code>
</pre>

**Anything else relevant?**

**Comment by aignas** [comment]
The correct way could be to start using techniques outlined in https://blog.engflow.com/2024/08/20/migrating-to-bazel-modules-aka-bzlmod---repo-names-and-rules_pkg/

**Comment by jkaye2012** [comment]
Sorry, I don't see how that has anything to do with the problem at hand. This should function without bzlmod and it's not a problem with pkg, but with python

**Comment by aignas** [comment]
Ah sorry, I might be missing something about the original problem. I was reading this blog post and the topics seemed to be related - packaging files.

**Comment by jkaye2012** [comment]
Nope. For now, we're just duplicating all of the dependencies within the wheel directly. An ugly workaround that we would prefer not to do, but nothing else that we've tried has worked.

I can't shake the feeling that we're doing something wrong, but I also don't know what that would be.

**Comment by gkouretas-maven** [comment]
Had the same issue and was able to ~fix~ workaround it by adding the external package to the `requires` argument of `py_wheel`:

For the example above:
```
py_wheel(
    name = "regression_harness_wheel",
    distribution = "regression_harness",
    version = "1.0.0",
    deps = [
        ":regression_harness_pkg",
    ],
    requires = [
        "behave",
    ],
)
```

**Comment by jkaye2012** [comment]
Yeah, that is the duplication of all dependencies within the wheel directly. It definitely works.

I wouldn't call that a fix though - when you have many dependencies themselves with potentially transitive third-party dependencies, it becomes a real pain (and it clearly something that _should_ be working, but the site-packages being in the wrong location prevents it from functioning properly).

**Comment by aignas** [comment]
Just to repeat what has been said in previous threads on Slack or GH issues somewhere.

`rules_python` accepts a lock file that is a list of packages to be installed to run tests, etc. 

Conceptually it is not the same as the list of compatible dependencies for some given `py_library` that you would like to package in a `py_wheel`.

In `py_wheel` we have no idea how to convert the dependencies to `Requires-Dist` that is the ask here. I am inclined to close this as won't do, because there is no way to get the `Requires-Dist` to be perfectly correct here. What we could do though is fail the `py_wheel` if there are missing requires-dist entries for the third-party transitive dependencies.

**Comment by rickeylev** [comment]
To add to what aignas said:

The lockfile has specific version specifications (e.g. `foo==2.0`), plus qualifiers for platform etc. The original specification (e.g. `foo > 1.0` from pyproject.toml) isn't part of the lockfile (nor is there a place to put them, in requirements.txt at least), so it can't be recovered.

At best, a generic "requires foo" could be generated by keeping track of what distributions are "direct" dependencies (this feels hand wavey, though).

What would probably make more sense is for py_wheel to accept a (potentially generated) file of requires-dist information. This would allow e.g. converting pyproject.toml into requires-dist entries, or using custom build rules to generate such content (e.g. an aspect with custom logic).

**Comment by aignas** [comment]
I guess the default of just including package distribution names in the default `Requires-Dist` would be good, because that is the minimum. If there are any more specific `Requires-Dist`, we could accept a file and-or statically supplied list.

The way we could do this is:
* from the `py_library` deps, collect all of the `METADATA` files.
* extract the `Name` from each `METADATA` file and write it as `Requires-Dist`.

This would probably not be too difficult (I could be convinced that we may need special providers in to make this easier because the METADATA files are special at the end of the day).

**Comment by jkaye2012** [comment]
I don't understand the internals of py_wheel, so please bear with me here. First a few assertions (please point out where any of these are incorrect):

* Bazel has knowledge of exact requirement versions from `pip_parse` using a lockfile
* Each individual target has knowledge of its direct requirements (we know this because it's only transitive dependencies that don't work, direct dependencies work normally)
* We can see the correct transitive dependencies being installed, they're just put in the wrong directory

I'm certain that I'm missing something here, because to me it seems like we have all of the information that we need to solve this problem - though to be fair, I don't know how to wire that up myself (otherwise, I would've already submitted the PR 😁). Maybe there is another workflow that doesn't using a lockfile that has to be supported?

> The lockfile has specific version specifications (e.g. foo==2.0), plus qualifiers for platform etc. The original specification (e.g. foo > 1.0 from pyproject.toml) isn't part of the lockfile (nor is there a place to put them, in requirements.txt at least), so it can't be recovered.

What does pyproject.toml have to do with anything here? Are you referring to the definition of the third-party dependency? None of the Bazel targets have a specification in this way, so I think that must be what you mean, but want to confirm.

**Comment by gkouretas-maven** [comment]
> Yeah, that is the duplication of all dependencies within the wheel directly. It definitely works.
> 
> I wouldn't call that a fix though - when you have many dependencies themselves with potentially transitive third-party dependencies, it becomes a real pain (and it clearly something that _should_ be working, but the site-packages being in the wrong location prevents it from functioning properly).

Agreed - definitely is not a sustainable thing, edited my comment to say “workaround”, since you are right that it technically is not a fix to the underlying issue…

**Comment by groodt** [comment]
There seem to be 2 distinct issues being discussed here?

1. `py_wheel` is "vendoring" / including transitive third-party dependencies in a nested site-packages folder in the built wheel distribution. It should generally only include the direct listed packages (in-tree code is an open question?) and not transitive third-party packages. These should be instead represented by `Requires-dist` metadata.
2. The distinction between requirement specifiers (eg `requests >= 1.0.0`) and concrete versions (eg `requests == 2.35.2`). When distributing a wheel to an index (eg PyPI or other) it is a best-practice to distribute the wheel with wide ranges of requirement specifiers (eg a lower-bound `requests >= 1.0.0` or even unbound `requests`). This is to ensure maximum compatibility for the resolver / installer that will be pulling down the wheel to install into their environment (Which may not be bazel. Statistically it won't be bazel). Distributing a wheel with `==` in `Requires-dist` effectively means that you are forcing anyone using your wheel to be using precisely the same environment as the publisher and will create dependency-hell on the consumption side. @aignas suggestion of including the "bare" name in `Requires-dist` is the most reasonable behaviour because it ensures a wide and compatible range for the installer. It does get a bit more complex than the name, because there are environment markers and other considerations to include based on the platform. Beyond that, as @rickeylev mentions, a more sophisticated approach of pulling the specifiers from `pyproject.toml` (or whatever source produced the lockfile) is an option.

Which problem does this issue relate to? I feel like it's the first one?

**Comment by jkaye2012** [comment]
@groodt I think you are correct that there are (at least) two technical issues, but from the user's perspective (or at least my perspective - I don't think I speak for everyone) the problem is simply:

> I specify dependencies in Bazel. When used as a build target, all dependencies are correctly resolved and use (e.g. by py_library). When used as a wheel, third-party transitive dependencies are not included at all, which leads the wheel to break

It's at the very least surprising behavior, and I would say against how we would _want_ py_wheel to work. I do understand the distinction that you're drawing from an implementation perspective, though.

**Comment by groodt** [comment]
Yes, so that's item 1. It is indeed surprising (and very likely against python packaging specs, but I'd need to dig a few things up) to create a "fat wheel" that vendors dependencies in this way. Vendoring things into "fat wheels" is a thing that can be done, but it is typically done by rewriting imports to not pollute the namespace and is very likely not what a normal user would be expecting.

I think the behaviour you are wanting, is that `py_wheel` will only include the explicitly listed in-tree code. This is the behaviour I would expect as well, since it's the same behaviour when using non-bazel wheel-building tools. 

Question for you: would you expect it to pick-up transitive in-tree code? eg There is a py_library `:foo` that depends on another in-tree py_library `:bar`. Lets assume no third-party dependencies from a lock-file for now. If I have a py_wheel that depends on `:foo`, should it include the code for `:bar`? Or require explicit packages to be listed? Different non-bazel build backends typically have configuration and auto-discovery for this.

eg. in `setuptools` it looks like this: https://setuptools.pypa.io/en/latest/userguide/package_discovery.html
```
# ...
[tool.setuptools]
packages = ["mypkg", "mypkg.subpkg1", "mypkg.subpkg2"]
# ...
```

**Comment by rickeylev** [comment]
I'd have to double check to make sure, but I, too, seem to recall py_wheel acting weird/surprising as to what code it includes. I think whether the py_package rule is used in combination also factors in (I think one takes all transitives by default, while the other doesn't, or something? And one has a "package" filter setting). A couple coworkers went through similar.

> Question for you: would you expect it to pick-up transitive in-tree code?

My take: Yes, I want it to. Why? Because otherwise you have to manually curate a separate set of targets for what gets packaged into the wheel. At best, this means having to list _every_ py_library in your project in the wheel, which is really annoying as targets change. It gets more complicated if you have a C module -- now you need to put a cc target, but a cc target doesn't have any python metadata (e.g. imports attribute), so relocating it to the right place can be a PITA

ps @groodt send me an email? Check any commit and my email will be there. Would be nice to stay in touch :)
